### DISCUSSION TITLE
[Plugin development] Performance bottleneck

### DISCUSSION THREAD

**Discussion by HananoshikaYomaru** [HEmile/juggl#115] [General]
# Thanks ✨

Thank you for this wonderful plugin. I have heard of this plugin when I first explored Obsidian community plugin, and recently I have to explore new graph view option because the internal graph view is already not enough for my need. I am currently using this plugin to show some clusters in my vault and it works well so far. 

# Performance bottleneck 

I expect my vault will eventually become very large because I am using a mono vault strategy and graph performance is my top concern. 

I know the original concept of Juggl does not support large graph but I want to know the performance bottleneck such that I can probably help develop it some day if I have time. Can someone explain a bit? 

 @HEmile

**Comment by HEmile** [comment]
Hi @HananoshikaYomaru , thank you for this discussion, and it's certainly a good one to have. Also thanks for possibly offering to help.

The situation is currently as follows. 
- I used pretty expensive edges in previous versions of Juggl (Bezier curves), that are pretty but known to not be very performative. I've recently changed those to straight edges which should be a lot quicker, but I have not tested this thorougly
    - I was pretty new to JS when I made this plugin, so I haven't done any other performance optimizations yet. Probably there's quite a bit to gain by just analyzing performance records and removing what is less essential
    - This page will give additional performance considerations https://js.cytoscape.org/#performance
- Regardless of anything, Juggl _will never be as quick as Obsidians internal graph_ 
    1. Obsidian uses a huge amount of internal optimization (like low-level, graphics code!) to make things super quick
    2. Juggl has _way_ more styling options and other features which give a lot of overhead
    3. The library Obsidian uses is a lot more low-level than Cytoscape.js, which means less overhead, but also makes it much harder to code things and extend the graph (hence why there are almost no plugins extending it). 
- I have huge vaults (two of both 10.000+ notes). Most definitely the global graph is completely unusable here.
    - However, local graphs are no problem at all!
    - That's why Juggl is designed as a tool to easily show _subgraphs_, and not the global graph. Imho, showing more than 1000 notes is just for nice visuals anyways... There are better algorithms out there to identify clusters (see for example our Graph Analysis plugin: https://github.com/SkepticMystic/graph-analysis) 
    - Hence why I chose not to focus on the global graph. I think Obsidians graph is easier to make look pretty too for large graphs.

**Reply by HananoshikaYomaru** [reply]
Thanks for reply. It is true that the occasion using global graph is few when the vault is large. But if I have time, I will probably help it out on optimisation. Does  https://github.com/SkepticMystic/graph-analysis integrate with BC and juggl? Can one simply visualize the cluster by a simple click in graph analysis?

**Reply by HEmile** [reply]
There is no integration, but that could be done fairly easily by looking at the breadcrumbs code and how it uses juggl
